src/adapters/MorphoMarketV1Adapter.sol

```diff
@@ -95,21 +100,17 @@ contract MorphoMarketV1Adapter is IMorphoMarketV1Adapter {
         require(msg.sender == parentVault, NotAuthorized());
         require(marketParams.loanToken == asset, LoanAssetMismatch());
 
-        if (assets > 0) {
-            (, uint256 redeemedShares) = IMorpho(morpho).withdraw(marketParams, assets, 0, address(this), address(this));
-            shares[marketId] -= redeemedShares;
-        }
-        // TODO
-        int256 change = int256(expectedSupplyAssets(marketParams, shares[marketId])) - int256(allocation(marketParams));
-
-        if (shares[marketId] == 0 && assets > 0) {
-            for (uint256 i = 0; i < allMarketParams.length; i++) {
-                if (Id.unwrap(allMarketParams[i].id()) == Id.unwrap(marketId)) {
-                    allMarketParams[i] = allMarketParams[allMarketParams.length - 1];
-                    allMarketParams.pop();
-                    break;
-                }
-            }
+        if (assets > 0) IMorpho(morpho).withdraw(marketParams, assets, 0, address(this), address(this));
+        // TODO: justify safe casting
+        int256 change =
+            int256(IMorpho(morpho).expectedSupplyAssets(marketParams, address(this))) - int256(allocation(marketParams));
+
+        if (indexInListPlusOne[marketId] != 0 && IMorpho(morpho).supplyShares(marketId, address(this)) == 0) {
+            MarketParams memory lastMarketParams = allMarketParams[allMarketParams.length - 1];
+            allMarketParams[indexInListPlusOne[marketId] - 1] = lastMarketParams;
```

Comment:
Also by moving from shares to asses, if the previous allocation was > 0 then the market is in the list (it is added only if the allocation goes > 0). So there is no need to store `index+1`.

see the [PR for 6](https://github.com/morpho-org/vault-v2/pull/658/files#diff-4898d1bb2c624323ad3314430022d7a2f6e5cdf59bc515ce07554072655ce524R113)